Move `Select` to `googlesitekit-components`

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #6653 

## Relevant technical choices

This PR moves the `Select` and `Option` components to the `googlesitekit-components` package.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

kuasha420 commented on assets/js/components/dashboard-sharing/DashboardSharingSettings/Module.js:
```diff
@@ -39,10 +39,9 @@ import {
  * Internal dependencies
  */
 import Data from 'googlesitekit-data';
-import { Tooltip } from 'googlesitekit-components';
+import { Tooltip, Select } from 'googlesitekit-components';
```
This is not a big deal, but should we sort the imports here? ie. `googlesitekit-components` before `googlesitekit-data` or `Select. Tooltip` instead of `Tooltip, Select`. Maybe this is something we should enforce project-wide and fix in a new issue using something like [eslint-sort-imports](https://eslint.org/docs/latest/rules/sort-imports)

nfmohit replied:
Updated all the imports in this PR, also opened #7275 to enforce it project-wide. Thank you!

kuasha420 replied:
Thank you for updating the imports here and creating the issue for fixing this project-wide!

kuasha420 commented on assets/js/googlesitekit/components-gm2/Option.js:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Option component.
```
We don't need this purely wrapper component and this will probably cause issues with `ref`. Instead, we should do what we did for the `Dialog` and it's associate components in the [components-gm2/index.js](https://github.com/google/site-kit-wp/blob/5a4f569143f54fc77117af6003839908ab84fab7/assets/js/googlesitekit/components-gm2/index.js#L19-L23).

nfmohit replied:
Updated, thanks!

kuasha420 commented on assets/js/googlesitekit/components-gm2/Select.js:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Select component.
```
Same here, see above comment.

nfmohit replied:
Updated, thanks!

kuasha420 reviewed (changes requested):
@nfmohit Thank you for creating the PR. I have left a few comment for you. Kindly have a look. Cheers!

nfmohit commented:
> @nfmohit Thank you for creating the PR. I have left a few comment for you. Kindly have a look. Cheers!

All addressed, back to you for another pass, thank you @kuasha420!

kuasha420 reviewed (approved):
@nfmohit Thank you! LGTM :1st_place_medal:

tofumatt reviewed (changes requested):
![CleanShot 2023-07-10 at 15 39 23](https://github.com/google/site-kit-wp/assets/90871/afe79048-00cf-4243-9eda-6cacf6df2335)

Looks like the `Select` component doesn't provide a default ID-to-label mapping. I noticed this when checking the storybook stories, because there the labels aren't connected semantically to the `select`.

I'd advocate adding this in this issue, using a similar approach with `useInstanceID` that we used fort the `TextField` component or similar. It's not obvious to a developer that the ID is REQUIRED to make this `select` accessible, and I actually didn't pick it up during accessibility reviews as we must be supplying IDs correctly elsewhere, or maybe I just missed it 🤔 

------

Other than that this looks good.

nfmohit commented:
> Looks like the `Select` component doesn't provide a default ID-to-label mapping. I noticed this when checking the storybook stories, because there the labels aren't connected semantically to the `select`.
> 
> I'd advocate adding this in this issue, using a similar approach with `useInstanceID` that we used fort the `TextField` component or similar. It's not obvious to a developer that the ID is REQUIRED to make this `select` accessible, and I actually didn't pick it up during accessibility reviews as we must be supplying IDs correctly elsewhere, or maybe I just missed it 🤔
> 
> Other than that this looks good.

As discussed in Slack, unfortunately, it looks like Material's `Select` component’s `enhanced` variant doesn’t pass the `id` prop to the `input` component:

https://github.com/material-components/material-components-web-react/blob/7c90f62e4b7c57dad1bc5dc01e7cd5772eddbe28/packages/select/EnhancedSelect.tsx#L183-L188C11

Regardless, I've implemented the `useInstanceID` approach that we used for the `TextField` component as it works for the non-`enhanced` variant. However, that isn't really necessary because there isn't a `<label>` element in that case so there's no need for an ID-to-label mapping. I've also added a non-`enhanced` basic variant of the component in Storybook for easier experimentation.

I think enforcing an ID on the hidden input wouldn't be very straight-forward, hence, we can probably address that in a newer issue with further investigation as everything else here is according to spec. What do you think?

Thank you, @tofumatt!

tofumatt reviewed (approved):
Indeed, looks like this is an issue with the Material select that's out-of-scope for the issue.

Thanks for the info here @nfmohit.
